Feature: Add Auth middleware and Authenticator.

kokokuo opened the pull request:
## Description
Support to provide built-in basic authenticator and authenticator abstract class for user customizing.

The **Authenticator** provides the **authentication** feature to make our user ( DA/DE ) could define the access way (e.g: username/password or token or credential ) and read each DC user's role with some customized attribute ( e.g: department ) for **authorization** used.

For the implementation, Vulcan provides three built-in authenticators and sets each authenticator option under `options` of `auth`.

### BasicAuthenticator
The HTTP Basic authenticator uses the authorization header `Basic <credential>` (insensitive for auth type) to authenticate. The credential encode base64 from `<username>:<password>`, and it supports two set user credentials method defined in project YAML.

```yaml
auth: 
  options: 
    // identifier for basic auth
    basic:
      htpasswd-file:
         path: <file-path>
         users:
           - name: user1
              attr:
                role: 'engineer' 
 
      users-list: 
        - name: user3
           md5Password: <md5-password>
           attr: 
             role: 'engineer'
```
The above `htpasswd-file` could set the user and md5 hash password in the file, `user-list` could let you set in yaml directly.

After you authenticate failed by request with authorization header `Basic <credential>`, then it will return the result with `WWW-Authenticate` header in the response.

### SimpleTokenAuthenticator
The Simple Token authenticator uses the authorization header `Simple-Token <credential>` to authenticate. The credential could be of any value and it authenticates from your user credentials in project YAML. 

```yaml
auth: 
  options: 
     // identifier for basic auth
    simple-token:
     - name: user3
        token: <any-value>
         attr: 
           role: 'engineer'
```
The above `token` in config is also could be any value, so the simple token authenticator is just a simple way to authenticate the same credentials from the authorization header `Simple-Token <credential>` (insensitive for auth type). In `SimpleTokenAuthenticator`, it will respond a JSON failed result like below when you authenticate failed:

```json
{
   "type": "simple-token",
  "message": "<the error message>"
}
```

### PasswordFileAuthenticator
The Password File authenticator uses the authorization header `Password-File <credential>` (insensitive for auth type) to authenticate. The credential encodes base64 from `<username>:<password>`, and it authenticates by the password file, we could set the path in project YAML.

```yaml
auth: 
  options: 
    // identifier for password-file auth
    password-file:
       path: <file-path>
       users:
         - name: user1
            attr:
               role: 'engineer' 
```
The password file should be `<username>:<bcrypt-password>` format in each line, it means the password need to encode `bcrypt` with round 10 first and then add the file. If not, we will show an error message to the notice user.

If authenticate failed, it will respond json format like the below:

```json
{
   "type": "password-file",
  "message": "<the error message>"
}
```


## How to use it
You can use our built-in authenticator and define the property with its identifier under `options` of `auth` module name in project config YAML to get the auth default defined information.  You could define one to multiple authenticator options in YAML depend on what authentication you want.

Please make sure to keep the same `attr` for consistency when you define the same user data in each different authenticator, but you could also omit this consistency when you want.

```yaml
auth: 
  options: 
     basic: 
       ...
     simple-token: 
       ...
     password-file: 
       ...
```

## Sample: doing the auth by the basic authenticator
- **Step1:** Add a user eason by using basic authentication with a checked token:
```yaml
auth: 
     options:
      basic:
        users-list: 
          - name: eason
             md5Password: '098f6bcd4621d373cade4e832627b4f6'
          attr:                 
             role: admin
```
- **Step2:** Send an API request with the user's basic auth token and passed
```yaml
Authorization: Basic ZWFzb246MDk4ZjZiY2Q0NjIxZDM3M2NhZGU0ZTgzMjYyN2I0ZjY=
```
- **Step3:** Passed, do the authorization in SQL file 
```sql
{% if context.user.attr.role != 'admin' %}
	error "No permission to do the query."
{% endif %}

select * from orders
```

## ⚠️  Notice 
The new version authenticator which connected which client-server is still designing, so I will create a new branch and pr to develop it after the new design confirms in the next meeting.


## How To Test / Expected Results
For the test result, please see the below test cases that passed the unit test:

<img width="740" alt="螢幕快照 2022-08-25 下午6 21 59" src="https://user-images.githubusercontent.com/5389253/186640560-64e6489b-5d0d-484d-b44f-25bd7f3bc68e.png">

## Commit Message
* caac5ee feat(serve): add auth middleware and authenticator
* 51444cb feat(serve): add authenticator and auth middleware test cases
* 898f0db feat(serve): add cursor, offset, keyset pgination strategy test cases.
* ba1b91f fix(serve): make authenticator and auth middleware have different result status.
* 8e9c780 fix(serve): refactor http basic and add simple-token, password-file authenticator.
* a1c8572 refactor(serve): add http basic authenticator
* c95709b fix(serve): update authenticator and auth middleware test cases

kokokuo commented:
The rebase has done from #23 which based on #25

oscar60310 commented on packages/serve/src/lib/auth/basicAuthenticator.ts:
```diff
@@ -0,0 +1,99 @@
+        break;
+      }
+    }
+
```
We can return here if no ansToken found:
```ts
if (ansToken === '') return {authenticated: false};
```

kokokuo replied:
Thanks for suggesting, I have fixed it!

oscar60310 reviewed (commented):
I have some suggestions about auth design. Thanks for adding test and factorying!

kokokuo commented on packages/serve/src/lib/auth/basicAuthenticator.ts:
```diff
@@ -0,0 +1,99 @@
+      }
+    }
+
+    if (srcToken !== ansToken) return { authenticated: false };
```
Thanks for reviewing and suggesting! 

I will change to [Trino](https://trino.io/docs/current/security/password-file.html) suggesting a password format ( bcrypt or PBKDF2 ), currently, seems bcrypt will be great for checking, I think.

And for the question, `Does it mean users need to set the correctly base64 encoded password to config?` > Yes

kokokuo replied:
Thanks for suggesting the hash password idea, I have adopted it in `BasicAuthenticator` and `PasswordFileAuthenticator`, each using a different hash way.

For the `SimpleTokenAuthetnciator`, I make it really easy for matching the value is same or not.

kokokuo commented on packages/serve/src/lib/auth/basicAuthenticator.ts:
```diff
@@ -0,0 +1,99 @@
+  private async verifyToken(srcToken: string, userOptions: UserAuthOptions) {
+    let ansToken = '';
+    const pattern = /^{{([\w]+|[ \w ]+)}}$/;
+    const matched = pattern.exec(userOptions.auth['token'] as string);
```
Thanks for @oscar60310 suggesting, according to our discussion, I will create the other PR to add it, currently, in this PR I remove environment variable part

kokokuo commented on packages/serve/src/lib/middleware/authMiddleware.ts:
```diff
@@ -0,0 +1,66 @@
+
+export interface AuthOptions {
+  ['user-auth']?: Array<UserAuthOptions>;
+}
```
Thanks for @oscar60310 reviewing and suggesting!  I make each authenticator option defined by identifier under the `options` of `auth`.

```yaml
auth: 
  options: 
     basic: 
       ...  // basic authenticator options
     simple-token: 
       ...  // simple-tokenauthenticator  options
     password-file: 
       ...  // password-file authenticator options
```

So like OIDC, oauth2 could also put under the `options` of `auth`. 

Thanks alot !

kokokuo commented on packages/serve/src/lib/middleware/authMiddleware.ts:
```diff
@@ -0,0 +1,66 @@
+    // authenticate each user by selected auth method in config
+    for (const name of Object.keys(this.authenticators)) {
+      const result = await this.authenticators[name].authenticate(
+        options['user-auth'] || [],
```
Thanks for reviewing and suggesting, after I adapt your suggestion for initializing in `onActivate` method, I won't pass the `options` into each `authenticate` method. Therefore, also no need to inject options.

kokokuo commented on packages/serve/src/lib/auth/basicAuthenticator.ts:
```diff
@@ -0,0 +1,99 @@
+
+    if (!matched) ansToken = ansToken || (userOptions.auth['token'] as string);
+    // matched[1] is env variable
+    if (matched) ansToken = ansToken || (process.env[matched[1]] as string);
```
Thanks for catching the part, I have removed it.

kokokuo commented on packages/serve/src/models/extensions/authenticator.ts:
```diff
@@ -0,0 +1,34 @@
+  ): Promise<AuthResult>;
+}
+
+@VulcanExtension(TYPES.Extension_Authenticator)
```
Thanks for suggesting, I have fixed the part to add it.

kokokuo commented on packages/serve/src/lib/auth/basicAuthenticator.ts:
```diff
@@ -0,0 +1,99 @@
+        const result = await this.verifyTokenInFile(token, userOptions);
+        if (result.authenticated) return result;
+      }
+    }
```
Thanks for reviewing and suggesting, it's a great suggestion, I have fixed to add the read options in `onActivate`, thanks !

kokokuo commented on packages/serve/src/lib/middleware/authMiddleware.ts:
```diff
@@ -0,0 +1,66 @@
+      await next();
+      return;
+    }
+    throw new Error('authentication failed.');
```
Thanks for suggesting, that after I add the multiple auth status for distinguishing what is authentication successful, or failed and "skip to the next authenticator" ( The status called `incorrect` now )  for checking. 

Now I have added the `www-authenticate` header in response when the authriozation request is basic and options have set basic options data.

kokokuo commented:
Hi @oscar60310, I have fixed all parts from your comment and suggestion!  I have also updated the PR description.

Thanks a lot for your reviewing and suggesting!

oscar60310 commented on packages/serve/src/models/extensions/authenticator.ts:
```diff
@@ -0,0 +1,52 @@
+   */
+  SUCCESS = 'SUCCESS',
+  FAIL = 'FAIL',
+  INCORRECT = 'INCORRECT',
```
NIT: Can we use other words like "indeterminate", "skipped", "bypass" instead of "incorrect"? "Incorrect" makes me think about "bad credentials", and "fail" means failure to authenticate due to internal issues.

kokokuo replied:
Thanks for the suggestion! I chose the `indeterminate` for renaming,

oscar60310 commented on packages/serve/src/lib/middleware/authMiddleware.ts:
```diff
@@ -0,0 +1,79 @@
+
+    this.authenticators = authenticators.reduce<AuthenticatorMap>(
+      (prev, authenticator) => {
+        if (authenticator.activate) authenticator.activate();
```
Activate functions are asynchronous, we can't wait for them to fulfill in the constructor, maybe we need a activate function for middleware too, and call it from upstream.

kokokuo replied:
Thanks for suggestion, I have refactored it, please check :D

oscar60310 reviewed (approved):
LGTM

kokokuo commented:
Hi @oscar60310 I have fixed the remaining two-part, please check it, thanks so much!
